### PR TITLE
Add CompoundLayout

### PR DESCRIPTION
**Pull request by luigiberrettini**
As per issue #1543 I added a CompoundLayout Layout to make it possible to use multiple nested heterogeneous layouts and create complex log entries.

Take as an example the following configuration:
```xml
<nlog>
  <targets>
    <target name='file' type='File' fileName='log.txt'>
      <layout type='CompoundLayout'>
        <layout type='SimpleLayout' text="myAmazingText: " />
        <layout type='JsonLayout'>
          <attribute name='time' layout='${longdate}' />
          <attribute name='level' layout='${level:upperCase=true}'/>
          <attribute name='nested' encode='false'  >
            <layout type='JsonLayout'>
              <attribute name='message' layout='${message}' />
              <attribute name='exception' layout='${exception}' />
            </layout>
          </attribute>
        </layout>
      </layout>
    </target>
  </targets>
  <rules>
  </rules>
</nlog>
```

Log entries would be like this:
```json
myAmazingText: { "time": "2016-10-30 13:30:55.0000", "level": "INFO", "nested": { "message": "this is message", "exception": "test" } }
```

### CODE REVIEW

**Comment by 304NotModified** [comment]
\0/

File: src/NLog/Config/XmlLoggingConfiguration.cs [hunk 1]
```diff
@@ -1102,6 +1104,15 @@ private bool AddArrayItemFromElement(object o, NLogXmlElement element)
             return false;
         }
 
+        private object CreateArrayItemInstance(NLogXmlElement element, Type elementType)
```
Comment (304NotModified) [inline src/NLog/Config/XmlLoggingConfiguration.cs]: I think this is OK, but maybe nice to add some (XML) comments.

File: src/NLog/Layouts/CompoundLayout.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+
+namespace NLog.Layouts
+{
+    using Config;
+    using System.Collections.Generic;
+    using System.Text;
+
+    /// <summary>
+    /// A layout containig a nested layouts.
```
Comment (304NotModified) [inline src/NLog/Layouts/CompoundLayout.cs]: Proposal: "A layout containing one or more nested layouts."

File: src/NLog/Layouts/CompoundLayout.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CompoundLayout"/> class.
+        /// </summary>
+        public CompoundLayout()
+        {
+            Layouts = new List<Layout>();
+        }
+
+        /// <summary>
+        /// Gets the array of inner layouts.
```
Comment (304NotModified) [inline src/NLog/Layouts/CompoundLayout.cs]: Proposal: remove the word "array" as it's a implementation detail

File: src/NLog/Layouts/CompoundLayout.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+{
+    using Config;
+    using System.Collections.Generic;
+    using System.Text;
+
+    /// <summary>
+    /// A layout containig a nested layouts.
+    /// </summary>
+    [Layout("CompoundLayout")]
+    [ThreadAgnostic]
```
Comment (304NotModified) [inline src/NLog/Layouts/CompoundLayout.cs]: I don't think this is true? It's depended of the layouts used?

File: src/NLog/Layouts/CompoundLayout.cs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    [AppDomainFixedOutput]
```
Comment (304NotModified) [inline src/NLog/Layouts/CompoundLayout.cs]: Same for this

File: src/NLog/Layouts/CompoundLayout.cs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        /// <param name="logEvent">The log event to be formatted.</param>
+        /// <returns>A string representation of the log event.</returns>
+        protected override string GetFormattedMessage(LogEventInfo logEvent)
+        {
+            var sb = new StringBuilder();
+            foreach (var layout in Layouts)
+                sb.Append(layout.Render(logEvent));
+            return sb.ToString();
+        }
+    }
```
Comment (304NotModified) [inline src/NLog/Layouts/CompoundLayout.cs]: I think we need also to implement `InitializeLayout` ?

**Comment by 304NotModified** [comment]
Looks good, added some notes!

File: src/NLog/Layouts/CompoundLayout.cs (hunk 4 above)
Comment (luigiberrettini) [inline src/NLog/Layouts/CompoundLayout.cs]: Should be A layout containing nested/inner layouts (nested to it not layouts with nested layouts)

File: src/NLog/Layouts/CompoundLayout.cs (hunk 6 above)
Comment (luigiberrettini) [inline src/NLog/Layouts/CompoundLayout.cs]: I initially implemented InitializeLayout but later removed it:

1. CsvLayout implements it but only to set properties that ar not Lists

2. JsonLayout contains a set of attributes and does not implement it... similarly CompoundLayout contains a set of Layouts

3. When I call render each Layout is initialized

I can implement both method as a loop over Layouts that calls them for each layout

File: src/NLog/Layouts/CompoundLayout.cs (hunk 5 above)
Comment (luigiberrettini) [inline src/NLog/Layouts/CompoundLayout.cs]: You are right I only looked at two layouts who share these attributes.
Do you think it is enough to remove the two attributes or should I, in some way, programmatically determine if all layouts have an attribute and set it?

Comment (304NotModified) [inline src/NLog/Layouts/CompoundLayout.cs]: yes then can be removed

**Comment by 304NotModified** [comment]
Any idea if we could test this case:

image

(`CreateArrayItemInstance` returning `null`)
See https://codecov.io/gh/NLog/NLog/compare/698ca65c2ea8ac508a13f8d49d0e1c3c7a012c5c...f6d0d4a92c528cfd97927222847ac8b81ad0de87

**Comment by 304NotModified** [comment]
I think we almost ready to merge this. 

Can you add those on the wiki? Public is fine. Please add some examples and "introduced in NLog 4.3.6"

**Comment by luigiberrettini** [comment]
We can add a test for layout whose XML configuration does not have a type: in this case I think an exception will be thrown.
This test does not depend on the CompoundLayout

**Comment by luigiberrettini** [comment]
I just added the documentation.

https://github.com/NLog/NLog/wiki/Layouts
(I suppose the NuGet badge will auto-update to 4.3.6 when you release it)

https://github.com/NLog/NLog/wiki/CompoundLayout

**Comment by 304NotModified** [comment]
Great! Thanks!

**Comment by 304NotModified** [comment]
It's online!

**Pull request merged**